// nostr-signer-capacitor-plugin/dist/esm/index.js
import { Capacitor, registerPlugin } from '@capacitor/core';
const native = registerPlugin('NostrSignerPlugin');
const ANDROID_ONLY = 'ANDROID_ONLY';
const ensureAndroid = () => {
    if (Capacitor.getPlatform() !== 'android') {
        const err = new Error('nostr-capacitor is Android-only. Use a NIP-07 provider on Web/iOS.');
        // @ts-expect-error attach code
        err.code = ANDROID_ONLY;
        throw err;
    }
};
export const buildPermissionsJson = (perms) => JSON.stringify(perms !== null && perms !== void 0 ? perms : []);
function normalizePermissions(permissions) {
    if (permissions == null)
        return undefined;
    if (typeof permissions === 'string')
        return permissions;
    return buildPermissionsJson(permissions);
}
export const NostrSignerPlugin = {
    async setPackageName(packageName) {
        ensureAndroid();
        if (!packageName)
            throw new Error('MISSING_PARAMS: packageName');
        await native.setPackageName({ packageName });
    },
    async isExternalSignerInstalled(packageName) {
        ensureAndroid();
        return native.isExternalSignerInstalled({ packageName });
    },
    async getInstalledSignerApps() {
        ensureAndroid();
        return native.getInstalledSignerApps();
    },
    async getPublicKey(packageName, permissions) {
        ensureAndroid();
        const perm = normalizePermissions(permissions);
        return native.getPublicKey({ packageName, permissions: perm });
    },
    async signEvent(packageName, eventJson, id, npub) {
        ensureAndroid();
        if (!eventJson || !id || !npub) {
            throw new Error('MISSING_PARAMS: eventJson,id,npub');
        }
        return native.signEvent({ packageName, eventJson, eventId: id, npub });
    },
    async nip04Encrypt(packageName, plainText, id, pubKey, npub) {
        ensureAndroid();
        if (!plainText || !pubKey || !npub) {
            throw new Error('MISSING_PARAMS: plainText,pubKey,npub');
        }
        return native.nip04Encrypt({ packageName, plainText, pubKey, npub, id });
    },
    async nip04Decrypt(packageName, encryptedText, id, pubKey, npub) {
        ensureAndroid();
        if (!encryptedText || !pubKey || !npub) {
            throw new Error('MISSING_PARAMS: encryptedText,pubKey,npub');
        }
        return native.nip04Decrypt({ packageName, encryptedText, pubKey, npub, id });
    },
    async nip44Encrypt(packageName, plainText, id, pubKey, npub) {
        ensureAndroid();
        if (!plainText || !pubKey || !npub) {
            throw new Error('MISSING_PARAMS: plainText,pubKey,npub');
        }
        return native.nip44Encrypt({ packageName, plainText, pubKey, npub, id });
    },
    async nip44Decrypt(packageName, encryptedText, id, pubKey, npub) {
        ensureAndroid();
        if (!encryptedText || !pubKey || !npub) {
            throw new Error('MISSING_PARAMS: encryptedText,pubKey,npub');
        }
        return native.nip44Decrypt({ packageName, encryptedText, pubKey, npub, id });
    },
    async decryptZapEvent(packageName, eventJson, id, npub) {
        ensureAndroid();
        if (!eventJson || !npub) {
            throw new Error('MISSING_PARAMS: eventJson,npub');
        }
        return native.decryptZapEvent({ packageName, eventJson, npub, id });
    },
    async getRelays(packageName, id, npub) {
        ensureAndroid();
        if (!id || !npub) {
            throw new Error('MISSING_PARAMS: id,npub');
        }
        return native.getRelays({ packageName, id, npub });
    },
};
